WIP: Introduce Colors plugin

igorlukanin opened the pull request:
This pull request provides reasonable default `borderColor` and `backgroundColor` if datasets don't have these properties set. Datasets will cycle through a list of 7 pre-defined "brand" colors (blue, red, orange, etc.).

Before:
```js
const data = {
  labels: [ ... ],
  datasets: [
    {
      label: 'Dataset 1',
      data: [ ... ],
      borderColor: '#36A2EB',
      backgroundColor: '#9BD0F5',
    },
    {
      label: 'Dataset 2',
      data: [ ... ],
      borderColor: '#FF6384',
      backgroundColor: '#FFB1C1',
    }
  ]
};
```

After:
```js
const data = {
  labels: [ ... ],
  datasets: [
    {
      label: 'Dataset 1',
      data: [ ... ],
    },
    {
      label: 'Dataset 2',
      data: [ ... ],
    }
  ]
};
```

It resolves #815, #884, and #1618 and makes Chart.js more developer-friendly, especially for those who are just getting started. It removes the need for excessive configuration and brings Chart.js closer to other charting libraries that also provide default colors.

As @etimberg recommended in [this comment](https://github.com/chartjs/Chart.js/issues/1618#issuecomment-229511158), it makes sense to have this as a plugin. When Chart.js is imported as `chart.js/auto`, the plugin will have an immediate effect. When Chart.js is imported as `chart.js`, the "default colors" behavior is opt-in—the plugin will have to be imported and registered:

```js
Chart.register(Colors);
```

This is work-in-progress, but I'd like to get feedback. The PR obviously lacks docs and maybe tests. Also, when it's merged, the excessive configuration would be removed from docs in a separate PR.

LeeLenaleee commented:
@kurkle already made a plugin for this which I think should be good enough, many we can link to that instead from the starter page of the docs 🤔 

https://github.com/kurkle/chartjs-plugin-autocolors

igorlukanin commented:
@LeeLenaleee Wow, thanks for pointing it out! I somehow missed that one 😅

I'm fully ready to abandon my code (haha, it's just a dozen lines of very trivial code). However, I think there's a lot of value in a plugin that is bundled with Chart.js. It would help drop the bar for those who are getting started.

What do you think about bundling @kurkle's plugin?

kurkle commented:
My plugin does not really generate brand colors, but if that is ok, its fine by me to bundle it. If it is bundled, it would probably be best to move it to the org.

kurkle commented:
And to this PR, it would also be fine by me to include a set of brand colors. I'd actually like that more, because of the brand.

stockiNail commented:
What about colorscheme plugin?

igorlukanin commented:
Currently working on an update to this PR. @LeeLenaleee @etimberg Can we mark this for inclusion to v4? I believe, it would be very appropriate.

igorlukanin commented:
@kurkle @LeeLenaleee @etimberg @dangreen I've rewritten the [docs page for Colors](http://localhost:8080/docs/master/general/colors.html) and incorporated the recommendations for using plugins there. Could you please take a look? Do you think it would work that way?

dangreen commented on docs/general/colors.md:
````diff
@@ -1,55 +1,139 @@
+All you need is to import and register the plugin:
+
+```javascript
+import { Colors } from `chart.js/plugins`;
+
+Chart.register(Colors);
````
```suggestion
import { Colors } from 'chart.js';

Chart.register(Colors);
```

dangreen commented on docs/general/colors.md:
````diff
@@ -1,55 +1,139 @@
 ```javascript
 const img = new Image();
 img.src = 'https://example.com/my_image.png';
 img.onload = function() {
````
nit

```suggestion
img.onload = () => {
```

dangreen reviewed (changes requested):
Did you test it with all chart types? Cause not all chart types to look good with borders.

Also for certain chart types, backgroundColor should be an array of colors, for example: https://www.chartjs.org/docs/latest/samples/other-charts/doughnut.html

LeeLenaleee commented on docs/general/colors.md:
```diff
@@ -1,55 +1,155 @@
+
+</div>
+
+You can also use [`chartjs-plugin-colorschemes`](https://github.com/nagix/chartjs-plugin-colorschemes) that comes with dozens of [pre-defined color series](https://nagix.github.io/chartjs-plugin-colorschemes/colorchart.html) from well-known tools like Tableau and Microsoft Office:
```
Currently colorschemes plugin only works with V2 so we either need to update it or dont list it

stockiNail replied:
I agree. When I mentioned `colorscheme` plugin in above https://github.com/chartjs/Chart.js/pull/10764#issuecomment-1272082436, I meant that I think the user's expectations will be to have something similar, embedded in chart.js, with similar features (i.e., custom color scheme, override the dataset colors configuration).

kurkle commented on docs/general/colors.md:
````diff
@@ -1,55 +1,155 @@
+
+```javascript
+Chart.defaults.backgroundColor = '#9BD0F5';
+Chart.defaults.borderColor = '#36A2EB';
````
Should add `color` to this list to be complete

kurkle reviewed (commented):
There is `chartjs-plugin-gradient` also, that could be mentioned here.
 
This comment is here only to give more reason for directing users to the awesome list instead of listing some of the available plugins here.

So, I'd remove the autocolors and colorscheme mentions and point users in that section to the awesome list to look for more options. Other than that, this looks awesome to me :)

stockiNail commented on src/plugins/plugin.colors.ts:
```diff
@@ -0,0 +1,105 @@
+];
+
+// Primary colors with 50% transparency
+const BACKGROUND_COLORS = /* #__PURE__ */ BORDER_COLORS.map(_ => _.replace(')', ', 0.5)'));
```
Could make sense to have an option for the alpha for `backgroundColor`, instead of hardcoded value?

igorlukanin replied:
@stockiNail Replied here: https://github.com/chartjs/Chart.js/pull/10764#discussion_r1000587010

stockiNail commented on src/plugins/plugin.colors.ts:
```diff
@@ -0,0 +1,105 @@
+  borderColor?: unknown;
+}
+
+const BORDER_COLORS = [
```
Could make sense to have an option to allow the user to use own colors scheme?

stockiNail replied:
Yes, but I think this will be requested in the future because the users prefer to have something out-of-the-box instead of having a dependency with other plugins. And having already done the 80% of the work, I think it will happen. Anyway, just my 2 cents.

igorlukanin replied:
@stockiNail Adding options for the background alpha and a user-defined color scheme make a lot of sense! However, I don't have an intention to "circumvent" the color-related options that Chart.js have or provide a way to "bypass" them. 

Here's my reasoning: if anyone wants to have a custom palette, why wouldn't them just set `borderColor`/`backgroundColor` on their datasets, as suggested in the docs? I think that if this plugin tries to "compete" with the existing color-related options, it will definitely fail. So, I don't want event try that 😄

stockiNail commented on src/plugins/plugin.colors.ts:
```diff
@@ -0,0 +1,105 @@
+      options: {elements},
+      data: {datasets}
+    } = chart.config;
+    let colorizer: DatasetColorizer;
```
This definition could be moved after the check if the colors are already set, because if the check is true, it's not used.

dangreen replied:
I guess, for modern js environments, it is not a (performance) problem.

stockiNail replied:
Not for a performance problem, but, at least for my "poor" mindset, I don't like to define a variable that I'm not sure to use.

igorlukanin commented:
@LeeLenaleee @kurkle @stockiNail @dangreen Thanks a lot for your comments and help!

@kurkle I've removed references to individual plugins and added a link to [awesome](https://github.com/chartjs/awesome). However, I think that the plugins in the awesome list can be rearranged for the ease of use, hence this PR: https://github.com/chartjs/awesome/pull/52

@LeeLenaleee I've pinged `colorschemes` author on [GitHub](https://github.com/nagix/chartjs-plugin-colorschemes/pull/34#issuecomment-1285469120) and [Twitter](https://twitter.com/igorlukanin/status/1583074591387176967). Hopefully, it would help bring this plugin back on track!

@stockiNail Thanks for great [suggestions](https://github.com/chartjs/Chart.js/pull/10764#discussion_r1000527850)! I hope my reasoning for not introducing them in this plugin make sense.

Looks like the updated docs page and the plugin is ready for the final review 😄

stockiNail commented:
> @stockiNail Thanks for great [suggestions](https://github.com/chartjs/Chart.js/pull/10764#discussion_r1000527850)! I hope my reasoning for not introducing them in this plugin make sense.

I didn't reply you but yes!! Thank you very much!! Great job, guys! I really appreciate it.

etimberg commented on src/plugins/plugin.colors.ts:
```diff
@@ -0,0 +1,106 @@
+];
+
+// Border colors with 50% transparency
+const BACKGROUND_COLORS = /* #__PURE__ */ BORDER_COLORS.map(color => color.replace('rgb(', 'rgba(').replace(')', ', 0.5)'));
```
might be able to avoid the `__PURE__` part by declaring `BORDER_COLORS` with `as const`

igorlukanin replied:
@etimberg I checked with @dangreen and he thinks it would make no difference for tree-shaking 🤷
